// jest-stare/lib/utils/IO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IO = void 0;
const fs = require("fs");
const path = require("path");
const pkgUp = require("pkg-up");
class IO {
    static unlinkSync(file) {
        if (IO.existsSync(file)) {
            fs.unlinkSync(file);
        }
    }
    static writeFile(wpath, data) {
        return new Promise((resolve, reject) => {
            fs.writeFile(wpath, data, (error) => {
                if (error) {
                    reject(error);
                }
                resolve();
            });
        });
    }
    static writeFileSync(wpath, data) {
        fs.writeFileSync(wpath, data);
    }
    static mkDirSync(dir) {
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir);
        }
    }
    static mkdirsSync(dir) {
        const dirs = path.resolve(dir).replace(/\\/g, "/").split("/");
        let createDir = "";
        for (const crDir of dirs) {
            createDir += (crDir + "/");
            IO.mkDirSync(createDir);
        }
    }
    static readFileSync(wpath) {
        return fs.readFileSync(wpath).toString();
    }
    static existsSync(wpath) {
        return fs.existsSync(wpath);
    }
    static readPackageJson() {
        const packageJson = pkgUp.sync();
        if (packageJson !== null) {
            return JSON.parse(IO.readFileSync(packageJson));
        }
        else {
            return {};
        }
    }
    static copyFileSync(src, dest) {
        fs.copyFileSync(src, dest);
    }
}
exports.IO = IO;
