// jest-stare/lib/render/diff/ImageSnapshotDifference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageSnapshotDifference = void 0;
const $ = require("jquery");
const Constants_1 = require("../../processor/Constants");
const AnsiParser = require("ansi-parser");
class ImageSnapshotDifference {
    static containsDiff(jestFailureMessage) {
        let isFailure = false;
        for (const indicator of ImageSnapshotDifference.DIFF_INDICATOR) {
            if (jestFailureMessage.indexOf(indicator) >= 0) {
                isFailure = true;
                break;
            }
        }
        return isFailure;
    }
    static generate(jestFailureMessage) {
        const imageDiffFilename = ImageSnapshotDifference.parseDiffImageName(jestFailureMessage);
        const errorMessage = ImageSnapshotDifference.parseDiffError(jestFailureMessage);
        const diffDiv = document.createElement("div");
        diffDiv.className = "image-snapshot-diff";
        const diffMessage = document.createElement("span");
        diffMessage.textContent = errorMessage;
        diffMessage.className = "text-muted";
        diffDiv.appendChild(diffMessage);
        const diffImageLink = document.createElement("a");
        diffImageLink.href = Constants_1.Constants.IMAGE_SNAPSHOT_DIFF_DIR + imageDiffFilename;
        diffDiv.appendChild(diffImageLink);
        const diffImage = document.createElement("img");
        diffImage.src = Constants_1.Constants.IMAGE_SNAPSHOT_DIFF_DIR + imageDiffFilename;
        diffImageLink.appendChild(diffImage);
        return $(diffDiv).get(0);
    }
    static parseDiffImagePath(jestFailureMessage) {
        const match = ImageSnapshotDifference.DIFF_IMAGE.exec(jestFailureMessage);
        if (match) {
            return AnsiParser.removeAnsi(match[1]).trim();
        }
        return null;
    }
    static parseDiffImageName(jestFailureMessage) {
        const path = ImageSnapshotDifference.parseDiffImagePath(jestFailureMessage);
        if (path) {
            return path.replace(/^.*[\\\/]/, "");
        }
    }
    static parseDiffError(jestFailureMessage) {
        const match = ImageSnapshotDifference.DIFF_DETAILS.exec(jestFailureMessage);
        if (match) {
            return match[1];
        }
        return null;
    }
}
exports.ImageSnapshotDifference = ImageSnapshotDifference;
ImageSnapshotDifference.DIFF_INDICATOR = ["different from snapshot", "image to be the same size"];
ImageSnapshotDifference.DIFF_IMAGE = /See diff for details:\s*((.*?)\.png)/;
ImageSnapshotDifference.DIFF_DETAILS = /Error: (.*)/;
