// jest-stare/lib/render/diff/TestDifference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestDifference = void 0;
const diff2html = require("diff2html");
const $ = require("jquery");
class TestDifference {
    static containsDiff(jestFailureMessage) {
        return jestFailureMessage.search(TestDifference.DIFF_INDICATOR) >= 0;
    }
    static generate(jestFailureMessage) {
        const jestDiff = TestDifference.isolateDiff(jestFailureMessage);
        const diffjson = diff2html.parse(jestDiff);
        const diffHtml = diff2html.html(diffjson, {
            drawFileList: false,
            outputFormat: "side-by-side",
            matching: "lines"
        });
        return $(diffHtml).get(0);
    }
    static isolateDiff(jestFailureMessage) {
        const beginIndex = jestFailureMessage.search(TestDifference.DIFF_INDICATOR);
        const endIndex = jestFailureMessage.search(TestDifference.DIFF_END_INDICATOR);
        let isolated = jestFailureMessage.substring(beginIndex, endIndex);
        let snapshotChanges = 0;
        let receivedChanges = 0;
        const changeLines = isolated.split(/\r?\n/g);
        for (const line of changeLines) {
            if (/^- /.test(line)) {
                snapshotChanges++;
            }
            else if (/^\+ /.test(line)) {
                receivedChanges++;
            }
        }
        const changesIndicator = `\n@@ -0,${snapshotChanges} +0,${receivedChanges} @@\n`;
        isolated = isolated.replace("- Snapshot", "--- Snapshot");
        isolated = isolated.replace("+ Received", "+++ Received");
        const lines = isolated.split(/\r?\n/g);
        lines.splice(2, 0, changesIndicator);
        return lines.join(`\n`);
    }
}
exports.TestDifference = TestDifference;
TestDifference.DIFF_INDICATOR = /- Snapshot\s*(\-\s*[0-9]+)?\n\s*\+ Received\s*(\+\s*[0-9]+)?/g;
TestDifference.DIFF_END_INDICATOR = /(at .*? \(.*?:[0-9]+:[0-9]+\)\s)/g;
